hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -14,8 +32,296 @@ import HIE.Bios as BIOS
+  ChComponentInfo ->
+  Bool
+partOfComponent fp comp
+  | Just normFp <- normalisedFp fp (ciSourceDirs comp), normFp `inTargets` getTargets comp fp = True
+  | otherwise = False
+  where
+    -- | Assuming a FilePath "src/Lib/Lib.hs" and a ciSourceDirs ["src"]
+    -- into 'Just "Lib"'
+    -- >>> normalisedFp "src/Lib/Lib.hs" ["src"]
+    -- Just "Lib/Lib.hs"
+    --
+    -- >>> normalisedFp "src/Lib/Lib.hs" ["app"]
+    -- Nothing
+    normalisedFp file sourceDirs = listToMaybe
+      $ mapMaybe ((`stripPrefix` file) . addTrailingPathSeparator) sourceDirs
+
+    inTargets :: FilePath -> [String] -> Bool
+    inTargets modFp targets =
+      -- Change a FilePath of the Form "Haskell/IDE/Engine/Cradle.hs" -> "Haskell.IDE.Engine.Cradle"
+      let modName = map
+            (\c -> if isPathSeparator c
+                   then '.'
+                   else c)
+            (dropExtension modFp)
+      in any (`elem` targets) [modName, fp]
+
+-- | Get the flags necessary to compile the given component.
+getFlags :: ChComponentInfo -> [String]
+getFlags = ciGhcOptions
+
+-- | Get all Targets of a Component, since we want to load all components.
+-- FilePath is needed for the special case that the Component is an Exe.
+-- The Exe contains a Path to the Main which is relative to some entry
+-- in 'ciSourceDirs'.
+-- We monkey-patch this by supplying the FilePath we want to load,
+-- which is part of this component, and select the 'ciSourceDir' we actually want.
+-- See the Documentation of 'ciSourceDir' to why this contains multiple entries.
+getTargets :: ChComponentInfo -> FilePath -> [String]
+getTargets comp fp = case ciEntrypoints comp of
+  ChSetupEntrypoint {} -> []
+  ChLibEntrypoint { chExposedModules, chOtherModules }
+    -> map unChModuleName (chExposedModules ++ chOtherModules)
+  ChExeEntrypoint { chMainIs, chOtherModules }
+    -> [sourceDir </> chMainIs | Just sourceDir <- [sourceDirs]]
+    ++ map unChModuleName chOtherModules
+    where
+      sourceDirs = find (`isFilePathPrefixOf` fp) (ciSourceDirs comp)
+
+-- | For all packages in a project, find the project the given FilePath
+-- belongs to most likely.
+findPackageFor :: NonEmpty (Package pt) -> FilePath -> Maybe (Package pt)
+findPackageFor packages fp = packages
+  & NonEmpty.toList
+  & sortOn (Down . pSourceDir)
+  & filter (\p -> pSourceDir p `isFilePathPrefixOf` fp)
+  & listToMaybe
+
+-- | Helper function to make sure that both FilePaths are normalised.
+-- Checks whether the first FilePath is a Prefix of the second FilePath.
+-- Intended usage:
+--
+-- >>> isFilePathPrefixOf "./src/" "./src/File.hs"
+-- True
+--
+-- >>> isFilePathPrefixOf "./src" "./src/File.hs"
+-- True
+--
+-- >>> isFilePathPrefixOf "./src/././" "./src/File.hs"
+-- True
+--
+-- >>> isFilePathPrefixOf "./src" "./src-dir/File.hs"
+-- True -- This is not really intended.
+isFilePathPrefixOf :: FilePath -> FilePath -> Bool
+isFilePathPrefixOf dir fp = normalise dir `isPrefixOf` normalise fp
```

Comment:
Not really sure that all these calls to `normalise` are necessary. There should be an invariant to store absolute paths somewhere, created by `canonicalisePath`.

Reply:
The implementation is just horribly wrong. Anything related to FilePaths has a high likelihood of being wrong :/